### ISSUE TITLE
🐞 PermissionError [WinError 32]

### ISSUE THREAD

**Issue by 4x0v7** [dagger/dagger#4172]
### What is the issue?

As reported first on Discord https://discord.com/channels/707636530424053791/1051860116175073292

When running `pipenv run python test.py` for the example python file the below message is thrown

```
File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\dagger\engine\docker.py", line 120, in start_sync
    engine_session_bin_path = tmp_bin_path.rename(engine_session_bin_path)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\scoop\apps\python311\current\Lib\pathlib.py", line 1175, in rename
    os.rename(self, target)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\winuser\\.cache\\dagger\\temp-dagger-engine-session-6r7_6_n7' -> 'C:\\Users\\winuser\\.cache\\dagger\\dagger-engine-session-9ad18aca71b1b1ba.exe'
```

If the file `temp-dagger-engine-session-6r7_6_n7`  is manually renamed to `dagger-engine-session-9ad18aca71b1b1ba.exe` and then `pipenv run python test.py` is re-run, dagger works successfully 

### Log output

```
PS> pipenv run python test.py
Traceback (most recent call last):
  File "C:\Users\winuser\git\dgrpy\test.py", line 29, in <module>
    anyio.run(test)
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\_core\_eventloop.py", line 70, in run
    return asynclib.run(func, *args, **backend_options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\_backends\_asyncio.py", line 292, in run
    return native_run(wrapper(), debug=debug)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\scoop\apps\python311\current\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\scoop\apps\python311\current\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\scoop\apps\python311\current\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\_backends\_asyncio.py", line 287, in wrapper      
    return await func(*args)
           ^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\git\dgrpy\test.py", line 13, in test
    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\dagger\connection.py", line 39, in __aenter__
    await self.engine.__aenter__()
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\dagger\engine\base.py", line 30, in __aenter__
    await self.start()
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\dagger\engine\bin.py", line 108, in start
    await anyio.to_thread.run_sync(self.start_sync)
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\_backends\_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\anyio\_backends\_asyncio.py", line 867, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\.virtualenvs\dgrpy-7tEWe_Su\Lib\site-packages\dagger\engine\docker.py", line 121, in start_sync       
    engine_session_bin_path = tmp_bin_path.rename(engine_session_bin_path)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\winuser\scoop\apps\python311\current\Lib\pathlib.py", line 1175, in rename
    os.rename(self, target)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\winuser\\.cache\\dagger\\temp-dagger-engine-session-6j7h5l2w' -> 'C:\\Users\\winuser\\.cache\\dagger\\dagger-engine-session-9ad18aca71b1b1ba.exe'
```

### Steps to reproduce

Install Python 3.10.9 or 3.11.1 on Windows using scoop

Create file `test.py` as per the getting started docs: https://docs.dagger.io/sdk/python/628797/get-started

`scoop install python311`
`pip install pipenv`
`pipenv install dagger-io` (should install v0.2.1 from PyPi)
`pipenv run python test.py`

I also tried without using `pipenv` and manually creating a venv. The issue persists.

After discussing on Discord with @marcosnils , adding a line above this line:
https://github.com/dagger/dagger/blob/sdk/python/v0.2.1/sdk/python/src/dagger/engine/docker.py#L117
with `tmp_bin.close()`
resolved the issue.

```diff
diff --git a/sdk/python/src/dagger/engine/docker.py b/sdk/python/src/dagger/engine/docker.py
index 549893f3..e2bd2234 100644
--- a/sdk/python/src/dagger/engine/docker.py
+++ b/sdk/python/src/dagger/engine/docker.py
@@ -114,6 +114,7 @@ class Engine(BinEngine):
                 finally:
                     cleanup()

+                tmp_bin.close()
                 tmp_bin_path = Path(tmp_bin.name)
                 tmp_bin_path.chmod(0o700)
```

De-denting these 3 lines also solves it:

```diff
diff --git a/sdk/python/src/dagger/engine/docker.py b/sdk/python/src/dagger/engine/docker.py
index 549893f3..5719f10c 100644
--- a/sdk/python/src/dagger/engine/docker.py
+++ b/sdk/python/src/dagger/engine/docker.py
@@ -114,10 +114,10 @@ class Engine(BinEngine):
                 finally:
                     cleanup()

-                tmp_bin_path = Path(tmp_bin.name)
-                tmp_bin_path.chmod(0o700)
+            tmp_bin_path = Path(tmp_bin.name)
+            tmp_bin_path.chmod(0o700)

-                engine_session_bin_path = tmp_bin_path.rename(engine_session_bin_path)
+            engine_session_bin_path = tmp_bin_path.rename(engine_session_bin_path)    

             # garbage collection of old engine_session binaries
             for bin in cache_dir.glob(f"{ENGINE_SESSION_BINARY_PREFIX}*"):
```

Running a trace using Sysinternals Process Monitor shows a `SHARING VIOLATION` event just after docker threads exit

![image](https://user-images.githubusercontent.com/19608884/207181753-85e45558-a868-439e-af6c-670a8afccf82.png)

### Dagger version

dagger 0.3.7 (2bd84ee24) windows/amd64

### OS version

Windows 10, Windows 11

**Comment by sipsma** [comment]
Thanks @4x0v7, I'm making some big changes to all this code anyways, but the file will still be downloaded (just from http rather than docker) so I'll be sure to include the close of the file in the right place. FWICT the other SDKs have the close line, it just got missed in python.

**Comment by helderco** [comment]
@sipsma this should be fixed in #4194 right?

**Comment by sipsma** [comment]
> @sipsma this should be fixed in https://github.com/dagger/dagger/pull/4194 right?

Ah yes it should be. There's another tweak needed to get windows working again (the windows cli releases use a zip instead of tar.gz), but just put that in the list of follow ups here: https://github.com/dagger/dagger/issues/3830

Closing this now
